### ISSUE TITLE
C++ codegen failure for derived objects; C backend works correctly

### ISSUE THREAD

**Issue by haxscramper** [nim-lang/Nim#16933]
Compiling derived `var` object on `C++` backend results in compilation
error. No problems on `C` backend.

### Example

```nim
type Base* = ref object of RootObj
method printSpace(c: var Base): void = discard

type Derived = ref object of Base
method printSpace(c: var Derived) = discard
```

### Current Output

Compiling code snippet above procedes normally on C backend, but fails on C++ compilation stage with following error:

```
error: cannot declare pointer to ‘struct tyObject_DerivedcolonObjectType___FW7aSxX39cPPnHcJYWsg0YA*&’
  181 | sF9bArgw9ctJODfybwVmujw((*((tyObject_DerivedcolonObjectType___FW7aSxX39cPPnHcJYWsg0YA*&*) (&(c)))));
```

### Expected Output

No compilation error for C++ backend

### Possible Solution

Offending cast is generated in `ccgtypes/getTypeDescAux()`.

https://github.com/nim-lang/Nim/blob/14c61ecf24110f3f6a1a4a9c88439e568065ddd9/compiler/ccgtypes.nim#L721 

I couldn't figure out how to correctly fix this - simply removing `cpp`
check obviously breaks a lot of other parts in compilation. If someone has
any ideas on fixing this I will try looking into this myself.

Generated calls for C++ backend differ only in `&` used

**c backend**

```c
printSpace__GsF9bArgw9ctJODfybwVmujw((*((
    tyObject_DerivedcolonObjectType___FW7aSxX39cPPnHcJYWsg0YA *
    /*{tfHasGCedMem, tfRefsAnonObj}tyRef*/* /*{tfHasGCedMem}tyVar*/*)(&(c)))));
```
 
**cpp backend** 
 
```cpp
printSpace__GsF9bArgw9ctJODfybwVmujw((*((
    tyObject_DerivedcolonObjectType___FW7aSxX39cPPnHcJYWsg0YA *
    /*{tfHasGCedMem, tfRefsAnonObj}tyRef*/& /*{tfHasGCedMem}tyVar*/*)(&(c)))));
```  

### Additional Information

I'g getting identical error when using `cligen` on CPP backend - I could
not minimize it due to lots of macro-generated code, but

```nim
import cligen
proc trailCompile*(file: string) = discard
dispatchMulti([trailCompile])
```

also fails. I'm not entirely sure if this is related, but it gives the same error message under the same conditions (`C++` dies, `C` ok)

```
$ nim -v
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2020-11-30
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 3fb5157ab1b666a5a5c34efde0f357a82d433d04
active boot switches: -d:release -d:danger
```

**Comment by ringabout** [comment]
It seems to be a regression. Works in 1.2.x

**Comment by ringabout** [comment]
I didn't bother to run a git bisect, but I skimmed all commits between 1.4.0 and 1.2.0. Fortunately I find the culprit at the second guess => https://github.com/nim-lang/Nim/pull/14161

**Comment by ringabout** [comment]
@github-actions nim cpp
```nim
type Base* = ref object of RootObj
method printSpace(c: var Base): void = discard

type Derived = ref object of Base
method printSpace(c: var Derived) = discard
```

**Comment by ringabout** [comment]
@juancarlospaco Hi, do you happen to know why the CI wasn't finished?

**Comment by juancarlospaco** [comment]
@ringabout The bug is a codegen bug, is generating:

```cpp
error: cannot declare pointer to ‘struct tyObject_DerivedcolonObjectType___QicCB7GN6dWokpZFgUI4JA*&’
  183 |                 printSpace__qGJls9cgdB4kneGWilyN1tg((*((tyObject_DerivedcolonObjectType___QicCB7GN6dWokpZFgUI4JA*&*) (&(c)))));
      |                                                                                                                   ^
Error: execution of an external compiler program 'g++ -c -std=gnu++14 -funsigned-char  -w -fmax-errors=3 -fpermissive   -I/home/runner/.choosenim/toolchains/nim-1.4.0/lib -I/home/runner/work/nimrun-action-minefield/nimrun-action-minefield -o /home/runner/work/nimrun-action-minefield/nimrun-action-minefield/@mtemp.nim.cpp.o /home/runner/work/nimrun-action-minefield/nimrun-action-minefield/@mtemp.nim.cpp' failed with exit code: 1
```

it is the line:

`printSpace__qGJls9cgdB4kneGWilyN1tg((*((tyObject_DerivedcolonObjectType___QicCB7GN6dWokpZFgUI4JA*&*) (&(c)))));`

and complaining on the `*&*`

The bot is failing because that sample is crashing all compilers from devel to 1.0
(older than 1.0 lang becomes too weird to bisect, strings and seqs are nil etc).

It needs at least 2 commits, 1 working, 1 non-working, 
then bisects between that range of commits until finds the breaking commit.

Basically that sample never worked properly in Nim.

That needs to be fixed/improved anyway, thanks for the info.

https://github.com/juancarlospaco/nimrun-action-minefield/actions/runs/5454609855/jobs/9924960902#step:4:166

**Comment by juancarlospaco** [comment]
@ringabout This commit introduced the bug:
https://github.com/nim-lang/Nim/commit/d6e1b5c07#diff-4509107d295d7d32b1887c8993cd0f56113ae60f36113e7d8778646dabd92ebcR2475

**Comment by ringabout** [comment]
!nim cpp

```nim
type Base* = ref object of RootObj
method printSpace(c: var Base): void = discard

type Derived = ref object of Base
method printSpace(c: var Derived) = discard
```

**Comment by ringabout** [comment]
@juancarlospaco Hello, would you mind introducing this super neat feature on the forum so that others can learn its existence?
